spice_agent/inference/actions.py

```diff
@@ -21,12 +23,156 @@
+                returned_embeddings_type=ReturnedEmbeddingsType.LAST_HIDDEN_STATES_NORMALIZED,
+                requires_pooled=False,
+                **DEFAULT_COMPEL_ARGUMENTS,
+            )
+
+        # Configure compel objects for StableDiffusionXL pipelines
+        if isinstance(pipe, StableDiffusionXLPipeline):
+            return Compel(
+                tokenizer=[pipe.tokenizer, pipe.tokenizer_2],
+                text_encoder=[pipe.text_encoder, pipe.text_encoder_2],
+                returned_embeddings_type=ReturnedEmbeddingsType.PENULTIMATE_HIDDEN_STATES_NON_NORMALIZED,
+                requires_pooled=[False, True],
+                **DEFAULT_COMPEL_ARGUMENTS,
+            )
+
+        if isinstance(pipe, StableDiffusionXLImg2ImgPipeline):
+            return Compel(
+                tokenizer=pipe.tokenizer_2,
+                text_encoder=pipe.text_encoder_2,
+                returned_embeddings_type=ReturnedEmbeddingsType.PENULTIMATE_HIDDEN_STATES_NON_NORMALIZED,
+                requires_pooled=True,
+                **DEFAULT_COMPEL_ARGUMENTS,
+            )
+
+        return None
+
+    def _run_compel(self) -> dict:
+        prompt_embeds = None
+        pooled_prompt_embeds = None
+        negative_prompt_embeds = None
+        negative_pooled_prompt_embeds = None
+
+        if self.compel.requires_pooled:
+            # Get pooled embeddings
+            prompt_embeds, pooled_prompt_embeds = self.compel(self.prompt)
+            negative_prompt_embeds, negative_pooled_prompt_embeds = self.compel(
+                self.negative_prompt
+            )
+        else:
+            prompt_embeds = self.compel(self.prompt)
+            negative_prompt_embeds = self.compel(self.negative_prompt)
+
+        # Pad prompt embeds
+        [
+            padded_prompt_embeds,
+            padded_negative_prompt_embeds,
+        ] = self.compel.pad_conditioning_tensors_to_same_length(
+            [prompt_embeds, negative_prompt_embeds]
+        )
+
+        embeddings = {
+            "prompt_embeds": padded_prompt_embeds,
+            "negative_prompt_embeds": padded_negative_prompt_embeds,
+        }
+
+        if pooled_prompt_embeds is not None:
+            embeddings["pooled_prompt_embeds"] = pooled_prompt_embeds
+
+        if negative_pooled_prompt_embeds is not None:
+            embeddings["negative_pooled_prompt_embeds"] = negative_pooled_prompt_embeds
+
+        return embeddings
+
+    def __call__(self) -> dict:
+        """
+        Returns prompt embeddings if a compel object exists; otherwise,
+        no embeddings are returned
+        """
+
+        LOGGER.info(""" [*] Generating embeddings. (Ignore token indices complaint)""")
+
+        self.without_embeddings = {
+            "prompt": self.prompt,
+            "negative_prompt": self.negative_prompt,
+        }
+
+        if self.compel:
+            try:
+                return self._run_compel()
+            except Exception as exception:
```

Comment:
maybe this should be a more explicit exception

Reply:
I really don't expect an exception to occur here so I still catch a generic exception to log it